src/zap/zapinfo.cpp

```diff
@@ -450,6 +450,24 @@ void ZapInfo::CompileMethod()
         return;
     }
 
+#if defined(_TARGET_X86_) || defined(_TARGET_AMD64_)
+    if (methodAttribs & CORINFO_FLG_JIT_INTRINSIC)
+    {
+        // Skip generating hardware intrinsic method bodies.
+        //
+        // The actual method bodies are only reachable via reflection/delegates, but we don't know what the
+        // implementation should do (whether it can do the actual intrinsic thing, or whether it should throw
+        // a PlatformNotSupportedException).
+
+        const char* namespaceName;
+        getMethodNameFromMetadata(m_currentMethodHandle, nullptr, &namespaceName, nullptr);
+        if (strcmp(namespaceName, "System.Runtime.Intrinsics.X86") == 0)
```

Comment:
Shouldn't this be anything under `System.Runtime.Intrinsics`? We also have the the `System.Runtime.Intrinsics.Arm` namespace (which should throw PNSE on x86) and the `Vector64/128/256<T>` types in the `System.Runtime.Intrinsics` namespace...

Reply:
Are they special?

https://github.com/dotnet/coreclr/blob/dd52a116323ff5deb78c8ad2f27481dc364ff590/src/vm/methodtablebuilder.cpp#L1508-L1518

This only sets the hardware intrinsic bit for `System.Runtime.Intrinsics.Arm.Arm64` on ARM64 and `System.Runtime.Intrinsics.X86` on xarch. Isn't stuff outside of that treated as regular IL?

Reply:
On x86 hardware, the arm ones are meant to throw PNSE and on arm hardware, the x86 ones should throw PNSE. At a minimum, I would expect an `#else if` clause handling ARM.

The logic you linked only applies to the "mass marking" of the methods as `Intrinsic`. We still have a couple of functions in the normal `System.Runtime.Intrinsics` namespace which are intrinsic as well (and are directly marked with the `Intrinsic` attribute).

For example, the following methods have software fallbacks but are also specially handled as intrinsics when the hardware supports it: https://github.com/dotnet/coreclr/blob/master/src/jit/hwintrinsiclistxarch.h#L30-L70. This allows us to generate better code (such as treating the call as a nop) when the hardware supports it.

Reply:
It would seem to me that we should always avoid generating method bodies for the intrinsics, regardless of target.